scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+        #                   with the row ids of all non-latest version file ids. These are the rows to soft delete.
+
+        # Note: The EXCEPT DISTINCT SELECT at the end grabs all row ids of rows that AREN'T the latest version.
+        # The final subquery here is in case there are multiple rows with the same version.
+        query = f"""
+        WITH allRows AS (SELECT datarepo_row_id AS rid, {target_table}_id AS fid, version AS v FROM {sql_table}),
+        latestFids AS (SELECT DISTINCT fid, MAX(v) AS maxv FROM allRows GROUP BY fid HAVING COUNT(1) > 1),
+        ridsOfAllFids AS (SELECT t.rid AS rid, f.fid, t.v FROM latestFids f JOIN allRows t ON t.fid = f.fid),
+        ridsOfLatestFids AS (SELECT t.rid AS rid, f.fid, f.maxv FROM latestFids f JOIN ridsOfAllFids t ON t.fid = f.fid AND t.v = f.maxv)
+        SELECT rid FROM ridsOfAllFids EXCEPT DISTINCT SELECT rid FROM (SELECT MAX(rid) AS rid, fid FROM ridsOfLatestFids GROUP BY fid)
+        """
+
+        return self._hit_bigquery(query)
+
+    def get_null_filerefs(self, target_table: str) -> Set[str]:
+        """
+        Determines what rows have null values in the file_id column. We want to soft delete those.
+        :param target_table: The particular table to operate on.
+        :return: A set of row ids to soft delete.
+        """
+        query = f"""
+        SELECT datarepo_row_id
+        FROM `{self.project}.datarepo_{self.dataset}.{target_table}` WHERE file_id IS NULL
+        """
+
+        return self._hit_bigquery(query)
+
+    def _hit_bigquery(self, query):
+        """
+        Helper function to consistently interact with biqquery while reusing the same client.
+        :param query: The SQL query to run.
+        :return: A set of whatever the query is asking for (assumes that we're only asking for a single column).
+        """
+        query_job = self.bigquery_client.query(query)
+        return {row[0] for row in query_job}
+
+    # local csv interactions
+    def create_csv(self, row_ids: Set[str], target_table: str) -> str:
+        """
+        Create a csv locally with one column filled with row ids to soft delete.
+        :param row_ids: A set of row ids to soft delete.
+        :param target_table: The table that the row ids belong to.
+        :return: The filename of the created csv.
+        """
+        filename = self.filename_template.format(table=target_table)
+        with open(filename, mode="w") as soft_delete_file:
+            sd_writer = csv.writer(soft_delete_file, delimiter=",", quotechar='"', quoting=csv.QUOTE_MINIMAL)
+
+            for rid in row_ids:
+                sd_writer.writerow([rid])
+
+        return filename
+
+    @staticmethod
+    def delete_csv(filename: str):
+        """
+        Remove the csv after it has been uploaded so that there isn't a deluge of csv files in the local directory.
+        :param filename: The filename of the file to delete.
+        :return:
+        """
+        os.remove(filename)
```

Comment:
consider using a temporary file in a context manager (e.g. `with TemporaryFile('filename.csv') as tempfile:`) instead of manually remembering to clean up after yourself

Reply:
(also, i'm thinking we can get away with replacing invocations of this function with just calling `os.remove` directly if we want to stay with this pattern)